### ISSUE TITLE
cumulative functions api design

### ISSUE THREAD

**Issue by semio** [semio/ddf_utils#25]
current format of accumulate:

``` yaml
procedure: accumulate
ingredients:
    - sg-datapoints
options:
    op:
        gdp: "aagr"  # run aagr() on gdp datapoints
```

but problem is:
1. functions like aagr() can be run in multiple timeframes, e.g n years growth rate where n is parameter.
2. we should only calculate these statistics within country, so we should always group them before accumulate.

To fix these problems we should add more options to the procedure, such as

``` yaml
procedure: accumulate
ingredients:
    - sg-datapoints
options:
    groupby:
        - geo
    op:
        gdp:
            func: aagr
            param:
                window: 10
```

EDIT: If we always need to groupby before accumulate. maybe combine `accumulate`  procedure into `groupby`?

**Comment by jheeffer** [comment]
I think there's no groupby for accumulative functions, just the column over which to accumulate. Then automatically you 'groupby' all the other key columns.

Starting data

| gender | geo | year | co2 |
|--------|-----|------|-----|
| male   | swe | 2015 | 5   |
| female | swe | 2015 | 7   |
| male   | swe | 2016 | 8   |
| female | swe | 2016 | 9   |
| male   | chn | 2014 | 4   |
| female | chn | 2014 | 3   |
| male   | chn | 2015 | 8   |
| female | chn | 2015 | 5   |

Accumulative sum over `year`

| gender | geo | year | co2 |
|--------|-----|------|-----|
| male   | swe | 2015 | 5   |
| female | swe | 2015 | 7   |
| male   | swe | 2016 | 13  |
| female | swe | 2016 | 18  |
| male   | chn | 2014 | 4   |
| female | chn | 2014 | 3   |
| male   | chn | 2015 | 12  |
| female | chn | 2015 | 8   |

You could even imagine a cumulative sum over `geo`:


I couldn't think of sensible way to do 'Cumulative sum over time by only geo'. The only thing that that could mean to me would be: first group by year and geo and then do cumulative sum over time.

So I think we can have the following configuration

```js
{
  procedure: 'accumulate',
  ingredients: 'sg-datapoints'
  result: 'sg-datapoints-accumulated'
  options: {
    accumulate_over: 'time',
    accumulate: {
      co2_emission: 'cumsum',
      income: {
        function: 'aagr',
        window: 10
      }
    }
  }
}
```
accumulate procedure can only be done on datapoints
accumulate_over must be of type time and be in key
accumulate must contain only value

accumulation

**Comment by jheeffer** [comment]
**Windowed function over multiple columns**

This is doable, following the logic described above. First sort by the accumulate_over columns, then sort by the other key columns. Form groups where the 'other key' columns are equal. Apply cumulative function per group.

In this case the order of columns matters. 

![image](https://cloud.githubusercontent.com/assets/1596071/19093410/61ad543a-8a8b-11e6-909f-104727ce4a32.png)

```
{
  procedure: 'accumulate',
  options: {
    accumulate_over: ['geo','time'] || ['time','geo'],
    accumulate: {
      co2: 'cumsum',
    }
  }
}
```
![image](https://cloud.githubusercontent.com/assets/1596071/19093387/35e88b9e-8a8b-11e6-9af6-06d068dcd52f.png)

However I have no clue when you would use something like this, so it doesn't need to be supported (for now).

**Comment by semio** [comment]
Thanks for the detail explanation!  l understand there are a few problems for the current api:

1. accumulate should be able to apply to any of the columns
1. There should be no `groupby` for `accumulate` function, we just first `groupby` and then apply accumulative function.
3. `accumulate` is too narrow, and `aarg` is not accumulative at all.

I agree your points and so I am thinking that if we can have a more general way to do applying functions to a ingredient. In pandas, there are a few ways to do this:

- dataframe.apply(function): apply function to each row or column
- dataframe.groupby.apply(function): apply function to each group
- dataframe.rolling.apply(function): apply function to each rolling windows
- and there are a few more, you can find it in the [doc](http://pandas.pydata.org/pandas-docs/stable/api.html#id5)

I think we can follow this setup and use the name `apply`

```javascript
{
  procedure: 'apply',
  ingredient: ['cdiac-datapoints']
  options: {
    apply_over: ['year','geo'] || ['geo','year'],
    function: {
      co2: 'cumsum'
    }
  }
}
```
 
As discussed in #4, we can return a grouped datapoint without aggregate. So we can run groupby first and apply a function to the grouped result:

```javascript
[
  {
    procedure: 'groupby',
    ingredient: ['cdiac-datapoints'],
    options: {
      groupby: 'geo'
    },
    result: 'cdiac-datapoints-grouped'
  },
  {
    procedure: 'apply',
    ingredient: ['cdiac-datapoints-grouped'],
    options: {
      apply_over: ['year','geo'] || ['geo','year'],
      function: {
        co2: 'cumsum'
      }
    }
  }
]
```

For windowed functions, we can treat it as a function that have a `window` parameter. I think we don't need an other `rolling` procedure, because the data return by rolling is meaningless without calculation.

```javascript
[
  {
    procedure: 'groupby',
    ingredient: ['cdiac-datapoints'],
    options: {
      groupby: 'geo'
    },
    result: 'cdiac-datapoints-grouped'
  },
  {
    procedure: 'apply',
    ingredient: ['cdiac-datapoints-grouped'],
    options: {
      apply_over: ['year','geo'] || ['geo','year'],
      function: {
        co2: 'aagr',
        window: 10
      }
    }
  }
]
```

Basically above is same as you recommended, but I just change the name to `apply`. Do you think it's OK?

**Comment by jheeffer** [comment]
I'll read into this a bit more, I think we really should try to follow pandas here, they though this out pretty well I think. Maybe we can support a subset of their window functions: http://pandas.pydata.org/pandas-docs/stable/api.html#window
Also use the same options as they do for their functions. It will make the mapping from recipe to pandas code a lot easier too I guess.

**Comment by jheeffer** [comment]
In general I have more and more the feeling we're creating a declarative layer on top of the pandas library, with some extra functions and input formats.

**Comment by jheeffer** [comment]
Okay, I dove into the pandas functions.

First I noticed in your examples above is that your conception of groupby is different than mine. Probably because of our backgrounds (me SQL and you python/pandas).

In pandas, running groupby on a dataframe returns a 'grouping' object. You can then apply different functions on that grouping. You can either transform, aggregate or filter per group.
In SQL, running groupby on a table returns a table. There is no transform or per group filter. The only possible function is aggregation. All columns in the result are either in group_by or aggregates functions.

I like the additional functionality of groupby in pandas though. However, we cannot have a grouping object be the output of a procedure. For simplicities sake, all procedures should always output ingredients (dataframes). Otherwise you're going into typing ingredients, which adds a new layer of complexity. 
So any groupby + function should be in one procedure I think.

Similar for window functions, both rolling windows and expanding windows. The definition of the window and the functions to apply should be in one procedure.

On the note of expanding windows: A cumsum is very similar to sum over expanding window, except for the handling of NaN's (see note in [this section](http://pandas.pydata.org/pandas-docs/stable/computation.html#id1)).

**Comment by jheeffer** [comment]
Now, let's define declarative procedures for these : )

**Comment by jheeffer** [comment]
One group by with aggregate, transform, filter option (only one allowed)
One window with aggregate option. Window size integer or time value is rolling, window size 'expanding' is expanding window.
Basic aggregate functions without grouping or window should fit under run_op I think?

**Comment by semio** [comment]
I see, thanks for making it clear :)

yes, it'd be good if you can write the specs, would make it easier for me to implement them :)

**Comment by jheeffer** [comment]
# groupby

```yaml
- procedure: groupby
  ingredient: population_and_foo_by_year_country_age_gender_education
  options:
    groupby: ["year","country"] || country
    aggregate:
      population: sum
      foo:
        function: bar
        param1: baz
    transform:
      population: foo
    filter:
      population: foo
  result: population_by_year_country
``` 

Only one of `aggregate`,`transform` or `filter` can be used in one procedure.
Any columns not mentioned in groupby or functions are dropped.

# window

```yaml
- procedure: window
  ingredient: immigration_surplus_by_year_country_gender
  options:
    window: 
      column: year # column which window is created from ("on" parameter in df.rolling)
      size: 5 || "expanding" # if positive integer or time offset, rolling window, if expanding, expanding window
      min_periods: 1 #optional, as in pandas
      center: false # optional, as in pandas
    aggregate:
      immigration_surplus: sum
      foo:
        function: bar
        param1: baz
  result: cumulative_immigration_surplus
``` 

- In multidimensional data, like datapoints, you want to apply the rolling function over groupings of the keys, except for the window column. In pandas implementation, you probably need some groupby to support multidimensional rolling: (not tested) `df.groupby(by=["country","gender"]).expanding(on="year").sum()`

**Comment by semio** [comment]
Looks good to me, I will try to implement them. 

For lambda functions, yes, security is the main concern. This is mainly because we need to use `eval()` to run the code, and `eval` can run any code provided. There are ways to limit the functions which can call by `eval`, but I think they are still not secure enough. See [here](http://nedbatchelder.com/blog/201206/eval_really_is_dangerous.html). 

I will dig into the lambda issue a bit later, see if there is a good way to do this.

**Comment by semio** [comment]
Because running python code in eval() is not safe and there is no other way to run python codes in recipe, I think we'd better just provide some pre-defined functions for recipe users. If we want the ability to create custom functions, we could make a function definition block in recipe and parse it. Just don't include python codes in the recipe.

**Comment by semio** [comment]
on window function: there is a bug using groupby with rolling on specific column for now, so I don't use the `on` parameter in rolling. (see the link in my commit message)

This don't have much impact on us because we usually have only 2 dimensions (geo and time), so if we group by geo then rolling will apply on time.

**Comment by jheeffer** [comment]
How does the script choose which column to roll on then?

**Comment by semio** [comment]
We choose by grouping by all keys other than the column to roll on. 

For example, the dataframe have index column ['geo', 'gender', 'year']

the column to roll is 'year'

then:

1. firstly group the dataframe by ['geo', 'gender'] (which is `['geo', 'year'].remove('year')`)
2. then each group have index 'year'
3. not using the `on` parma equals to rolling on index. In our case, because the index is 'year' column, so rolling on index is same as rolling on 'year' column
